mm/kfence/core.c

```diff
@@ -71,31 +74,135 @@ noinline void kfence_disable(void)
 	WRITE_ONCE(kfence_enabled, false);
 }
 
-/* TODO(glider): kernel_physical_mapping_change() is x86-only */
-unsigned long kernel_physical_mapping_change(unsigned long start, unsigned long end,
-					     unsigned long page_size_mask);
+static pgprot_t pgprot_clear_protnone_bits(pgprot_t prot)
+{
+	/*
+	 * _PAGE_GLOBAL means "global page" for present PTEs.
+	 * But, it is also used to indicate _PAGE_PROTNONE
+	 * for non-present PTEs.
+	 *
+	 * This ensures that a _PAGE_GLOBAL PTE going from
+	 * present to non-present is not confused as
+	 * _PAGE_PROTNONE.
+	 */
+	if (!(pgprot_val(prot) & _PAGE_PRESENT))
+		pgprot_val(prot) &= ~_PAGE_GLOBAL;
+
+	return prot;
+}
+
+/* Some code borrowed from arch/x86/mm/pat/set_memory.c. */
+static bool split_large_page(pte_t *kpte, unsigned long address, unsigned int level)
```

Comment:
This is all x86-specific, right? We're going to need to move this to something like '<asm/kfence.h>' eventually. A TODO would be good so we don't forget.

Reply:
I actually think page table interfaces are arch-agnostic, but the devil's in the details.
Let's keep a TODO around till we try it on ARM64.